core/blockchain_pruner.go

```diff
@@ -0,0 +1,256 @@
+			}
+			return err
+		}
+
+		bp.batchWriter = nil
+		return nil
+	} else {
+		return errors.New("end when batchWriter not init")
+	}
+}
+
+func (bp *blockchainPruner) runWhenNotPruning(cb func()) {
+	// check is start
+	if atomic.CompareAndSwapInt32(&bp.isPruneBeaconChainRunning, 0, 1) {
+		defer atomic.StoreInt32(&bp.isPruneBeaconChainRunning, 0)
+
+		cb()
+	}
+}
+
+func (bp *blockchainPruner) StartPruneBeaconChain(db ethdb.Database, maxBlockNum uint64, maxEpoch *big.Int) {
+	bp.runWhenNotPruning(func() {
+		err := bp.initPruneBeaconChain(db)
+		if err != nil {
+			utils.Logger().Info().Err(err).Msg("pruneBeaconChain init error")
+			return
+		}
+
+		defer func() {
+			err := bp.endPruneBeaconChain()
+			if err != nil {
+				utils.Logger().Info().Err(err).Msg("pruneBeaconChain end error")
+				return
+			}
+
+			utils.Logger().Info().
+				Uint64("maxBlockNum", maxBlockNum).
+				Uint64("maxEpoch", maxEpoch.Uint64()).
+				Int("deletedBlockCount", bp.deletedBlockCount).
+				Int("deletedValidatorSnapshot", bp.deletedValidatorSnapshot).
+				Int("skipValidatorSnapshot", bp.skipValidatorSnapshot).
+				Dur("cost", time.Now().Sub(bp.startTime)).
+				Msg("pruneBeaconChain delete block success")
+
+			if rand.Intn(compactProportion) == 1 {
+				startTime := time.Now()
+				// probability of 1 in 1000 blocks to Compact, It consumes a lot of IO
+				for _, rangeStartEnd := range bp.deleteRange {
+					err := db.Compact(rangeStartEnd[0], rangeStartEnd[1])
+					if err != nil {
+						if isUnrecoverableErr(err) {
+							fmt.Printf("Unrecoverable error when writing leveldb: %v\nExitting\n", err)
+							os.Exit(1)
+						}
+					}
+				}
+				utils.Logger().Info().
+					Uint64("maxBlockNum", maxBlockNum).
+					Uint64("maxEpoch", maxEpoch.Uint64()).
+					Dur("cost", time.Now().Sub(startTime)).
+					Msg("pruneBeaconChain compact db success")
+			}
+		}()
+
+		bp.pruneBeaconChainBlock(db, maxBlockNum)
+		bp.pruneBeaconChainValidatorSnapshot(db, maxEpoch)
+	})
+}
+
+func (bp *blockchainPruner) pruneBeaconChainBlock(reader ethdb.Database, maxBlockNum uint64) {
+	minKey, maxKey := rawdb.IteratorBlocks(reader, func(blockNum uint64, hash common.Hash) bool {
+		if blockNum >= maxBlockNum {
+			return false
+		}
+
+		// skip genesis block
+		if blockNum == 0 {
+			return true
+		}
+
+		blockInfo := rawdb.ReadBlock(reader, hash, blockNum)
+		err := bp.deleteBlockInfo(blockInfo)
+		if err != nil {
+			utils.Logger().Info().
+				Uint64("blockNum", blockNum).
+				AnErr("err", err).
+				Msg("pruneBeaconChain delete block info error")
+			return false
+		}
+
+		err = rawdb.DeleteBlock(bp, hash, blockNum)
+		if err != nil {
+			utils.Logger().Info().
+				Uint64("blockNum", blockNum).
+				AnErr("err", err).
+				Msg("pruneBeaconChain delete block error")
+			return false
+		}
+
+		// limit time spent
+		bp.deletedBlockCount++
+		return bp.deletedBlockCount < maxDeleteBlockOnce
+	})
+
+	bp.deleteRange = append(bp.deleteRange, [2][]byte{minKey, maxKey})
```

Comment:
I would personally suggest to return the slice instead of adding the result to structure. [Purely functional programming](https://en.wikipedia.org/wiki/Purely_functional_programming) shall be preferred over stateful ones.

Reply:
You are right, edited.